### DISCUSSION TITLE
Can I assume that `AesGcm.Encrypt/Decrypt` is safe for in-place operation?

### DISCUSSION THREAD

**Discussion by GF-Huang** [dotnet/runtime#95552] [Q&A] [answered]
I means that the plaintext and ciphertext buffer are same buffer.

Use case:
```cs
using var aes = new AesGcm(...);
var buffer = new byte[64];
var tag = new byte[16];
aes.Encrypt(
    nonce: ...,
    plaintext: buffer,
    ciphertext: buffer,
    tag: tag
);
```

And how can I determine if an algorithm is supports in-place operation?

**Comment by danmoseley** [comment]
@dotnet/area-System-Security

**Comment by bartonjs** [comment, marked as answer]
Yes, it's a thing we explicitly test for.  https://github.com/dotnet/runtime/blob/0c513d95c181159f3ea02531c7901ce15503f3ee/src/libraries/System.Security.Cryptography/tests/AesGcmTests.cs#L337-L354

I believe that all of our cryptographic routines allow the plaintext and ciphertext to be the same buffer.  My recollection was that we made "overlapping, but with an offset fail", but it has already been pointed out to me that sometimes we make that work.

**Reply by GF-Huang** [reply]
Thanks your point it out. Hopefully the documentation for the .NET API will mention this explicitly.

**Reply by danmoseley** [reply]
@GF-Huang could you open a docs issue if necessary?

**Reply by GF-Huang** [reply]
https://github.com/dotnet/dotnet-api-docs/issues/9494
